source/Halibut/Transport/Protocol/ControlMessageReader.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Halibut.Transport.Protocol
+{
+    internal class ControlMessageReader
+    {
+        internal string ReadUntilNonEmptyControlMessage(Stream stream)
+        {
+            while (true)
+            {
+                var line = ReadControlMessage(stream);
+                if (line.Length > 0) return line;
+            }
+        }
+
+        internal string ReadControlMessage(Stream stream)
+        {
+            StringBuilder sb = new StringBuilder();
+            while (true)
+            {
+                var nextByte = stream.ReadByte();
+                if (nextByte == -1) throw new EndOfStreamException();
+                // Control messages must end with \r\n and must not contain \r or \n
```

Comment:
I'm curious how is this ensured from the sender's side? `StreamWriter` seems to use `Environment.NewLine` as the line break.

Reply:
It looks like `\r\n` is always sent as part of https://github.com/OctopusDeploy/Halibut/blob/9fce5a8b3de869a3af7acaa0547172432326cae1/source/Halibut/Transport/Protocol/MessageExchangeStream.cs#L37